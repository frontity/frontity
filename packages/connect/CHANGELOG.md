# Change Log

## 1.3.0

### Minor Changes

- [`7111b3ce`](https://github.com/frontity/frontity/commit/7111b3cee7065816a885629cac93967f705d0969) [#415](https://github.com/frontity/frontity/pull/415) Thanks [@luisherranz](https://github.com/luisherranz)! - Migrate `@frontity/connect` to be a thin layer on top of `react-easy-state`. This means that in the future, taking advantage of the bug fixes and other improvements from `react-easy-state` will be much easier.

### Patch Changes

- [`641a1cf0`](https://github.com/frontity/frontity/commit/641a1cf00a02f78076e23b7918027d7219c08f58) [#791](https://github.com/frontity/frontity/pull/791) Thanks [@DAreRodz](https://github.com/DAreRodz)! - Modify derived props so they receive `state` and `libraries` as arguments. Also, fix the `Derived` type definition.

- Updated dependencies [[`641a1cf0`](https://github.com/frontity/frontity/commit/641a1cf00a02f78076e23b7918027d7219c08f58), [`86b2eff9`](https://github.com/frontity/frontity/commit/86b2eff993aac3e9360946a0c190e239b6f93abf)]:
  - @frontity/types@1.8.0

## 1.2.1

### Patch Changes

- [`b41175d0`](https://github.com/frontity/frontity/commit/b41175d0f5df9ca95fc449ca1a0eca6649f1bccf) [#683](https://github.com/frontity/frontity/pull/683) Thanks [@cristianbote](https://github.com/cristianbote)! - Reverts the preinstall hook added for development workflows.

- Updated dependencies [[`b41175d0`](https://github.com/frontity/frontity/commit/b41175d0f5df9ca95fc449ca1a0eca6649f1bccf)]:
  - @frontity/error@0.1.3

## 1.2.0

### Minor Changes

- [`3ced7fdf`](https://github.com/frontity/frontity/commit/3ced7fdfd93004c210bb47692ffae265874828e7) [#670](https://github.com/frontity/frontity/pull/670) Thanks [@nicholasio](https://github.com/nicholasio)! - Update react imports in all packages (including themes) to removed unnecessary React Imports

### Patch Changes

- [`01880c34`](https://github.com/frontity/frontity/commit/01880c34c111f55c23169adb7365ea9262e6cca8) [#666](https://github.com/frontity/frontity/pull/666) Thanks [@cristianbote](https://github.com/cristianbote)! - Adopt the new version of emotion 11. That means using the emotion deidcated babel plugin for rewriting the imports, generating source-maps and enable auto labels.

* [`5f329dab`](https://github.com/frontity/frontity/commit/5f329dabe9d67d0b3664938865491674ef798433) [#655](https://github.com/frontity/frontity/pull/655) Thanks [@mburridge](https://github.com/mburridge)! - Fix broken links in README files.

* Updated dependencies [[`5f329dab`](https://github.com/frontity/frontity/commit/5f329dabe9d67d0b3664938865491674ef798433)]:
  - @frontity/error@0.1.2

## 1.1.4

### Patch Changes

- [`d7b4b429`](https://github.com/frontity/frontity/commit/d7b4b429f1f23dfae74b9781ea1b1de00aed763c) [#616](https://github.com/frontity/frontity/pull/616) Thanks [@luisherranz](https://github.com/luisherranz)! - Stop using `instanceof Array` in favor of `Array.isArray()`.

* [`b766e330`](https://github.com/frontity/frontity/commit/b766e330465b6a76d927eaddaa763a684dc1b228) [#610](https://github.com/frontity/frontity/pull/610) Thanks [@michalczaplinski](https://github.com/michalczaplinski)! - Update the types so that state accepts parameters which are a union of derived state & some serializable value

## 1.1.3

### Patch Changes

- [`159e02ca`](https://github.com/frontity/frontity/commit/159e02ca080ec9f7004c90276621d1a2708192ce) [#508](https://github.com/frontity/frontity/pull/508) Thanks [@luisherranz](https://github.com/luisherranz)! - Fix the types of `connect` when imported from `frontity` and used with options, like:

  ```js
  import { connect } from "frontity";

  const Component = () = { /* Some React component... */ };

  export default connect(Component, { injectProps: false });
  ```

## 1.1.2

### Patch Changes

- [`1440cfe7`](https://github.com/frontity/frontity/commit/1440cfe77f1a9154562dff23a97185c77bebb59c) [#482](https://github.com/frontity/frontity/pull/482) Thanks [@juanmaguitar](https://github.com/juanmaguitar)! - README.md created for this package.

- Updated dependencies [[`fb67272b`](https://github.com/frontity/frontity/commit/fb67272bd8a3dfff00868af394484ec09f1e0785)]:
  - @frontity/error@0.1.1

## 1.1.1

### Patch Changes

- [`5cb29ab6`](https://github.com/frontity/frontity/commit/5cb29ab63ab31872a4d853e5e2fdbdabca974c9f) [#430](https://github.com/frontity/frontity/pull/430) Thanks [@luisherranz](https://github.com/luisherranz)! - Fix broken types in TypeScript 3.9.

## 1.1.0

### Minor Changes

- [`996865a2`](https://github.com/frontity/frontity/commit/996865a27690d5b89d2ef110f5b1bf3fb91da6f5) [#404](https://github.com/frontity/frontity/pull/404) Thanks [@orballo](https://github.com/orballo)! - Implements `useConnect()` and adds options object `{ injectProps: boolean }` to `connect()`.

## 1.0.5

### Patch Changes

- [`34b37ec8`](https://github.com/frontity/frontity/commit/34b37ec84bc4c387754f241ea56fd5eb8b13b14d) [#360](https://github.com/frontity/frontity/pull/360) Thanks [@luisherranz](https://github.com/luisherranz)! - Move `getSnapshot` from a store property to an import of `"@frontity/connect"`.

## 1.0.4

### Patch Changes

- [`696dec11`](https://github.com/frontity/frontity/commit/696dec11bb8d32f0821cca3f5ce39e27c42d60b6) [#296](https://github.com/frontity/frontity/pull/296) Thanks [@michalczaplinski](https://github.com/michalczaplinski)! - Fix a critical bug that causes promise errors to be swallowed

## [1.0.3](https://github.com/frontity/frontity/compare/@frontity/connect@1.0.2...@frontity/connect@1.0.3) (2019-10-02)

### Bug Fixes

- **typescript:** update to latest version ([a89b11c](https://github.com/frontity/frontity/commit/a89b11c))

## [1.0.2](https://github.com/frontity/frontity/compare/@frontity/connect@1.0.1...@frontity/connect@1.0.2) (2019-09-10)

### Bug Fixes

- **core:** remove componentWillMount warning with react-helmet-async ([0ea885b](https://github.com/frontity/frontity/commit/0ea885b))
- **wp-source:** properly populate custom post types and taxonomies ([857f803](https://github.com/frontity/frontity/commit/857f803))

## [1.0.1](https://github.com/frontity/frontity/compare/@frontity/connect@1.0.0...@frontity/connect@1.0.1) (2019-06-19)

### Bug Fixes

- **connect:** only run batchMethods if Proxy exists ([fa07e8e](https://github.com/frontity/frontity/commit/fa07e8e))

# [1.0.0](https://github.com/frontity/frontity/compare/@frontity/connect@0.1.4...@frontity/connect@1.0.0) (2019-06-05)

### Bug Fixes

- **all:** update typscript and fix some keywords ([1fe5fec](https://github.com/frontity/frontity/commit/1fe5fec))

### Features

- add some keywords ([9267faa](https://github.com/frontity/frontity/commit/9267faa))

### BREAKING CHANGES

- get this packages to 1.0.0

## [0.1.4](https://github.com/frontity/frontity/compare/@frontity/connect@0.1.3...@frontity/connect@0.1.4) (2019-06-05)

### Bug Fixes

- **connect:** include scheduler to fix batching in async events ([b4aa340](https://github.com/frontity/frontity/commit/b4aa340))

## [0.1.3](https://github.com/frontity/frontity/compare/@frontity/connect@0.1.2...@frontity/connect@0.1.3) (2019-05-17)

### Bug Fixes

- **connect:** support hooks inside connected components ([708dc95](https://github.com/frontity/frontity/commit/708dc95))

## [0.1.2](https://github.com/frontity/frontity/compare/@frontity/connect@0.1.1...@frontity/connect@0.1.2) (2019-05-15)

**Note:** Version bump only for package @frontity/connect

## 0.1.1 (2019-05-15)

**Note:** Version bump only for package @frontity/connect
