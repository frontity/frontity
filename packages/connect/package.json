{
  "name": "@frontity/connect",
  "version": "1.3.0",
  "description": "The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.",
  "keywords": [
    "frontity",
    "connect",
    "state-manager"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "main": "src/index.js",
  "types": "src/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "../../node_modules/.bin/jest --ci --coverage",
    "test": "../../node_modules/.bin/jest --watch",
    "test:inspect": "node --inspect ../../node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "@frontity/react-easy-state": "^7.0.0",
    "@frontity/observer-util": "^4.3.0",
    "@frontity/error": "^0.1.1",
    "@frontity/types": "^1.8.0"
  },
  "devDependencies": {
    "react-test-renderer": "^17.0.1"
  }
}
