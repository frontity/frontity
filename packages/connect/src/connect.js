import {
  useState,
  useEffect,
  useMemo,
  memo,
  useRef,
  createContext,
  useContext,
} from "react";
import { warn } from "@frontity/error";
import { observe, unobserve } from "@frontity/observer-util";

const COMPONENT = Symbol("owner component");

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext({});

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * A flag to know when a connected component is being rendered.
 */
let isConnected = false;

/**
 * The default options of connect.
 */
const defaultOptions = {
  /**
   * Define if `connect` should inject the store props in the component, or only
   * make it reactive. Useful when the store is accessed via `useConnect` and
   * the props are passed down an HTML node.
   */
  injectProps: true,
};

/**
 * Flag to know if there's been a change before the component was mounted.
 */
const changedBeforeMounted = new WeakMap();

/**
 * Flag to know when a component has been mounted.
 */
const mounted = new WeakMap();

/**
 * Connect a React component with the Frontity connect store and make it
 * reactive to its changes.
 *
 * @param Comp - The React component to be connected.
 * @param options - The connect options, defined in {@link defaultOptions}.
 *
 * @returns The same Component, but reactive and with the store included in the
 * props.
 */
export function connect(Comp, options) {
  const isFunctionalComp = !(Comp.prototype && Comp.prototype.isReactComponent);
  options = options ? { ...defaultOptions, ...options } : defaultOptions;

  let ReactiveComp;

  if (isFunctionalComp) {
    // Use a hook based reactive wrapper.
    ReactiveComp = (props) => {
      // Dummy setState to update the component.
      const [, setState] = useState();

      // Use a ref to store the reaction because we want to do some annotations.
      const reaction = useRef();

      // Create a memoized reactive wrapper of the original component (render)
      // at the very first run of the component function.
      const render = useMemo(
        () => {
          reaction.current = observe(Comp, {
            scheduler: () => {
              // Trigger a new rerender if the component has already been
              // mounted.
              if (mounted.get(reaction.current)) setState({});
              // Annotate it as "changed" if the component has not been mounted
              // yet.
              else changedBeforeMounted.set(reaction.current, true);
            },
            lazy: true,
          });

          // Initialize a flag to know if the component was finally mounted.
          mounted.set(reaction.current, false);

          // Initialize a flag to know if the was reaction was invalidated
          // before the component was mounted.
          changedBeforeMounted.set(reaction.current, false);

          return reaction.current;
        },

        // Adding the original Comp here is necessary to make HMR work. It does
        // not affect behavior otherwise.
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [Comp]
      );

      useEffect(() => {
        // Mark the component as mounted.
        mounted.set(reaction.current, true);

        // If there was a change before the component was mounted, trigger a
        // new rerender.
        if (changedBeforeMounted.get(reaction.current)) setState({});

        // Cleanup the reactive connections when the component is unmounted.
        return () => unobserve(reaction.current);
      }, []);

      // Extract the Frontity store from the context.
      const { state, actions, libraries } = useContext(context);

      // The isConnected flag is used to warn users that try to use `useConnect`
      // in a non-connected component.
      isConnected = true;
      try {
        // Run the reactive render instead of the original one and inject the
        // props if necessary.
        return render({
          ...props,
          ...(options.injectProps && state && { state }),
          ...(options.injectProps && actions && { actions }),
          ...(options.injectProps && libraries && { libraries }),
        });
      } finally {
        isConnected = false;
      }
    };
  } else {
    /**
     * Wrap the component to extract the store from the context and pass it
     * down to the final component. Also, make the render method reactive.
     */
    class ReactiveClassComp extends Comp {
      /**
       * Initialize the wrapper.
       *
       * @param props - The props passed to the original component.
       * @param context - The React context.
       */
      constructor(props, context) {
        super(props, context);

        // Create some state to be able to rerender the component later.
        this.state = this.state || {};
        this.state[COMPONENT] = this;

        // Make the render method reactive.
        this.render = observe(this.render, {
          scheduler: () => this.setState({}),
          lazy: true,
        });
      }

      /**
       * The render method of this React Class.
       *
       * @returns The same Component but with the store injected as props.
       */
      render() {
        const { state, actions, libraries } = this.context;
        const props = { ...this.props, state, actions, libraries };
        return <Comp {...props} />;
      }

      /**
       * React should trigger updates on prop changes, while  handles
       * store changes.
       *
       * @param nextProps - The props of the next render.
       * @param nextState - The state of the next render.
       * @returns Whether the component should rerender or not.
       */
      shouldComponentUpdate(nextProps, nextState) {
        const { props, state } = this;

        // Respect the case when the user defines a shouldComponentUpdate.
        if (super.shouldComponentUpdate) {
          return super.shouldComponentUpdate(nextProps, nextState);
        }

        // Return true if it is a reactive render or the state changes.
        if (state !== nextState) {
          return true;
        }

        // The component should also update if any of its props shallowly
        // changed value.
        const keys = Object.keys(props);
        const nextKeys = Object.keys(nextProps);
        return (
          nextKeys.length !== keys.length ||
          nextKeys.some((key) => props[key] !== nextProps[key])
        );
      }

      /**
       * Remove the reaction when the component is unmounted.
       */
      componentWillUnmount() {
        // Call user defined componentWillUnmount.
        if (super.componentWillUnmount) {
          super.componentWillUnmount();
        }
        // Clean up memory used.
        unobserve(this.render);
      }
    }

    // Attach the context and pass the compnent up.
    ReactiveClassComp.contextType = context;
    ReactiveComp = ReactiveClassComp;
  }

  // Make sure we display the correct name.
  ReactiveComp.displayName = Comp.displayName || Comp.name;

  // Static props are inherited by class components, but have to be copied for
  // function components.
  if (isFunctionalComp) {
    Object.keys(Comp).forEach((key) => {
      ReactiveComp[key] = Comp[key];
    });
  }

  return isFunctionalComp ? memo(ReactiveComp) : ReactiveComp;
}

/**
 * React hook that returns the Frontity store in connected components.
 *
 * @returns The Frontity store (state, actions, libraries...).
 */
export const useConnect = () => {
  if (!isConnected)
    warn(
      "Warning: useConnect() is being used in a non connected component, " +
        "therefore the component won't update on state changes. " +
        "Please wrap your component with connect().\n"
    );

  return useContext(context);
};
