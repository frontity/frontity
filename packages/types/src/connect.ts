/* eslint-disable jsdoc/require-jsdoc */

/**
 * TSDocs disabled by `@luisherranz` in this file because he needs to figure
 * out first which types will remain when we merge this PR
 * https://github.com/frontity/frontity/pull/415, taking into account that
 * `react-easy-state` and `@nx-js/observer-util` already have types themselves.
 */

import { ComponentProps, FunctionComponent, ComponentClass } from "react";
import Package from "./package";
import { ResolveState, ResolveActions, FilterInjectedProps } from "./utils";

export type Connect<Pkg extends Package, Props extends object = {}> = Omit<
  Pkg,
  "state" | "actions" | "name" | "roots" | "fills"
> & {
  state: ResolveState<Pkg["state"]>;
  actions: ResolveActions<Pkg["actions"]>;
} & Props;

export interface ConnectFunction {
  <Comp extends FunctionComponent<any>>(
    comp: Comp,
    options?: { injectProps: boolean }
  ): FunctionComponent<FilterInjectedProps<ComponentProps<Comp>>>;
  <Comp extends ComponentClass<any>>(
    comp: Comp,
    options?: { injectProps: boolean }
  ): ComponentClass<FilterInjectedProps<ComponentProps<Comp>>>;
}

export interface CreateStore {
  <Pkg extends Package>(pkg: Pkg): Omit<Pkg, "state" | "actions"> & {
    state: ResolveState<Pkg["state"]>;
    actions: ResolveActions<Pkg["actions"]>;
  };
}

export default Connect;
