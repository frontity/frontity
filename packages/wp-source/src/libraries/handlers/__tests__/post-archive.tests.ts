import { createStore, InitializedStore } from "@frontity/connect";
import wpSource from "../../../";
import WpSource from "../../../../types";
import Api from "../../api";
// JSON mocks
import { mockResponse } from "./mocks/helpers";
import posts from "./mocks/post-archive/posts.json";
import posts2 from "./mocks/post-archive/posts-page-2.json";
import postsCpt from "./mocks/post-archive/posts-cpt.json";

let store: InitializedStore<WpSource>;
let api: jest.Mocked<Api>;
beforeEach(() => {
  store = createStore(wpSource());
  store.actions.source.init();
  api = store.libraries.source.api as jest.Mocked<Api>;
});

describe("post archive", () => {
  test("works with first page", async () => {
    // Mock Api responses
    api.get = jest.fn().mockResolvedValueOnce(
      mockResponse(posts, {
        "X-WP-Total": "5",
        "X-WP-TotalPages": "2"
      })
    );
    // Fetch entities
    await store.actions.source.fetch("/");
    expect(store.state.source).toMatchSnapshot();
  });

  test("works with pagination", async () => {
    // Mock Api responses
    api.get = jest.fn().mockResolvedValueOnce(
      mockResponse(posts2, {
        "X-WP-Total": "5",
        "X-WP-TotalPages": "2"
      })
    );
    // Fetch entities
    await store.actions.source.fetch("/page/2/");
    expect(store.state.source).toMatchSnapshot();
  });

  test("fetchs from a different endpoint with extra params", async () => {
    // Add custom post endpoint and params
    store.state.source.postEndpoint = "multiple-post-type";
    store.state.source.params = { type: ["post", "cpt"] };
    // Mock Api responses
    api.get = jest.fn().mockResolvedValueOnce(
      mockResponse(postsCpt, {
        "X-WP-Total": "5",
        "X-WP-TotalPages": "2"
      })
    );
    // Fetch entities
    await store.actions.source.fetch("/");
    expect(api.get.mock.calls).toMatchSnapshot();
    expect(store.state.source).toMatchSnapshot();
  });
});
