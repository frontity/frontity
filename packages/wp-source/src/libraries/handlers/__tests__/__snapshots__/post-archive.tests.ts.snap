// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post archive fetchs from a different endpoint with extra params 1`] = `
Array [
  Array [
    Object {
      "endpoint": "multiple-post-type",
      "params": Object {
        "_embed": true,
        "page": 1,
        "search": undefined,
        "type": Array [
          "post",
          "cpt",
        ],
      },
    },
  ],
]
`;

exports[`post archive fetchs from a different endpoint with extra params 2`] = `
Object {
  "api": "https://test.frontity.io",
  "attachment": Object {
    "1": Object {
      "author": 1,
      "id": 1,
      "link": "/post-1/attachment-1/",
      "slug": "attachment-1",
      "type": "attachment",
    },
    "12": Object {
      "author": 1,
      "id": 12,
      "link": "/cpt/cpt-11/attachment-cpt-11/",
      "slug": "attachment-cpt-11",
      "type": "attachment",
    },
    "2": Object {
      "author": 1,
      "id": 2,
      "link": "/post-2/attachment-2/",
      "slug": "attachment-2",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "cpt": Object {
    "11": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          12,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 12,
      "id": 11,
      "link": "/cpt/cpt-11/",
      "slug": "cpt-11",
      "tags": Array [],
      "type": "cpt",
    },
  },
  "data": Object {
    "/": Object {
      "isArchive": true,
      "isFetching": false,
      "isHome": true,
      "isPostArchive": true,
      "isPostTypeArchive": true,
      "isReady": true,
      "items": Array [
        Object {
          "id": 1,
          "link": "/post-1/",
          "type": "post",
        },
        Object {
          "id": 11,
          "link": "/cpt/cpt-11/",
          "type": "cpt",
        },
        Object {
          "id": 2,
          "link": "/post-2/",
          "type": "post",
        },
      ],
      "total": 5,
      "totalPages": 2,
      "type": "post",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {
    "type": Array [
      "post",
      "cpt",
    ],
  },
  "post": Object {
    "1": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          1,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 1,
      "id": 1,
      "link": "/post-1/",
      "slug": "post-1",
      "tags": Array [],
      "type": "post",
    },
    "2": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          2,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 2,
      "id": 2,
      "link": "/post-2/",
      "slug": "post-2",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "multiple-post-type",
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
}
`;

exports[`post archive works with first page 1`] = `
Object {
  "api": "https://test.frontity.io",
  "attachment": Object {
    "1": Object {
      "author": 1,
      "id": 1,
      "link": "/post-1/attachment-1/",
      "slug": "attachment-1",
      "type": "attachment",
    },
    "2": Object {
      "author": 1,
      "id": 2,
      "link": "/post-2/attachment-2/",
      "slug": "attachment-2",
      "type": "attachment",
    },
    "3": Object {
      "author": 1,
      "id": 3,
      "link": "/post-3/attachment-3/",
      "slug": "attachment-3",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/": Object {
      "isArchive": true,
      "isFetching": false,
      "isHome": true,
      "isPostArchive": true,
      "isPostTypeArchive": true,
      "isReady": true,
      "items": Array [
        Object {
          "id": 1,
          "link": "/post-1/",
          "type": "post",
        },
        Object {
          "id": 2,
          "link": "/post-2/",
          "type": "post",
        },
        Object {
          "id": 3,
          "link": "/post-3/",
          "type": "post",
        },
      ],
      "total": 5,
      "totalPages": 2,
      "type": "post",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {
    "1": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          1,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 1,
      "id": 1,
      "link": "/post-1/",
      "slug": "post-1",
      "tags": Array [],
      "type": "post",
    },
    "2": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          2,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 2,
      "id": 2,
      "link": "/post-2/",
      "slug": "post-2",
      "tags": Array [],
      "type": "post",
    },
    "3": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          3,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 3,
      "id": 3,
      "link": "/post-3/",
      "slug": "post-3",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "posts",
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
}
`;

exports[`post archive works with pagination 1`] = `
Object {
  "api": "https://test.frontity.io",
  "attachment": Object {
    "4": Object {
      "author": 1,
      "id": 4,
      "link": "/post-4/attachment-4/",
      "slug": "attachment-4",
      "type": "attachment",
    },
    "5": Object {
      "author": 1,
      "id": 5,
      "link": "/post-5/attachment-5/",
      "slug": "attachment-5",
      "type": "attachment",
    },
  },
  "author": Object {
    "1": Object {
      "id": 1,
      "link": "/author/author-1/",
      "name": "Author 1",
      "slug": "author-1",
    },
  },
  "category": Object {
    "1": Object {
      "count": 5,
      "id": 1,
      "link": "/category/cat-1/",
      "parent": 0,
      "slug": "cat-1",
      "taxonomy": "category",
    },
  },
  "categoryBase": "",
  "data": Object {
    "/page/2/": Object {
      "isArchive": true,
      "isFetching": false,
      "isHome": true,
      "isPostArchive": true,
      "isPostTypeArchive": true,
      "isReady": true,
      "items": Array [
        Object {
          "id": 4,
          "link": "/post-4/",
          "type": "post",
        },
        Object {
          "id": 5,
          "link": "/post-5/",
          "type": "post",
        },
      ],
      "total": 5,
      "totalPages": 2,
      "type": "post",
    },
  },
  "get": [Function],
  "homepage": "",
  "isWpCom": false,
  "page": Object {},
  "params": Object {},
  "post": Object {
    "4": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          4,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 4,
      "id": 4,
      "link": "/post-4/",
      "slug": "post-4",
      "tags": Array [],
      "type": "post",
    },
    "5": Object {
      "_embedded": Object {
        "author": Array [
          1,
        ],
        "wp:featuredmedia": Array [
          5,
        ],
        "wp:term": Array [
          Array [
            1,
          ],
        ],
      },
      "author": 1,
      "categories": Array [
        1,
      ],
      "featured_media": 5,
      "id": 5,
      "link": "/post-5/",
      "slug": "post-5",
      "tags": Array [],
      "type": "post",
    },
  },
  "postEndpoint": "posts",
  "postsPage": "",
  "subdirectory": "",
  "tag": Object {},
  "tagBase": "",
}
`;
